### PR TITLE
Specify requested properties for Material Model

### PR DESCRIPTION
**Pull request by anne-glerum**
There are many places where we evaluate the material model to get at one material property. However, many other properties are also computed, like the reaction_terms. This PR specifies in several places that only the viscosity is requested. This should avoid computing the reaction terms and force terms for elasticity and strain dependent rheologies, as both Elasticity and StrainDependent check for `in.requests_property(MaterialProperties::reaction_terms)`.

However, currently request_property() still does
```
      //TODO: Remove this once all callers set requested_properties correctly
      if ((property & MaterialProperties::Property::reaction_terms) != 0)
        return (strain_rate.size() != 0);
```
so if in.strain_rate is filled (like for viscosity) then reaction_terms are still computed. 

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: source/postprocess/visualization/shear_stress.cc
Comment (gassmoeller) [inline source/postprocess/visualization/shear_stress.cc:61]: It would be nice if MaterialModelInputs had a constructor that would take this as input, then you could write above:
```
        MaterialModel::MaterialModelInputs<dim> in(input_data,
                                                   this->introspection(),
                                                   MaterialModel::MaterialProperties::viscosity);
```

Want to give this a try? If not the current form works as well.

**Review by gassmoeller — COMMENTED** [review]
As discussed this works and is a step in the right direction, but does not do anything at the moment because of the compatibility fixes in `MaterialModelInputs<dim>::requests_property`. Still a step forward to setting this correctly.

For the future: It would be nice if MaterialModelInputs had a constructor that would take this as input, then you could write:
```
        MaterialModel::MaterialModelInputs<dim> in(input_data,
                                                   this->introspection(),
                                                   MaterialModel::MaterialProperties::viscosity);
```

Do you want to give this a try? If not the current form works as well.

**Comment by bangerth** [comment]
On 7/16/21 10:50 AM, Rene Gassmoeller wrote:
> For the future: It would be nice if MaterialModelInputs had a constructor that 
> would take this as input, then you could write:
> 
> |MaterialModel::MaterialModelInputs<dim> in(input_data, this->introspection(), 
> MaterialModel::MaterialProperties::viscosity); |
> 
> Do you want to give this a try? If not the current form works as well.

Actually, I would prefer if this flag was set on the *output* object 
(MaterialModelOutputs). The *input* object does not need to know what it is 
being used for, but the output object should specify which fields need to be 
filled.

**Comment by anne-glerum** [comment]
@bangerth The requested_properties are already a part of MaterialModelInputs, are you saying that should be changed to MaterialModelOutputs? MaterialModels already check the input for the requested properties (although the check doesn't return the right answer for until it's implemented everywhere).

**Review by gassmoeller — APPROVED** [review]
Nice, thanks for the cleanup! I left a suggestion for how to resolve your question.

@bangerth: I see where you are coming from, but as @anne-glerum mentioned this is already established (#3457), this PR only uses the functionality in more places. I see why you say it could go into the outputs object, but I think when I implemented this I had the following two thoughts:
1. Telling the material model what to do is an input operation.
2. Having this in the output object would mean it is handed over as non-const, meaning the material model could modify what properties to compute internally (not likely, but it seems conceptually not ok).

If you want it to be moved I don't feel very strongly about it, but my preference would be an input. A fix would only need a bit of search and replace.

In any case I think we should not delay this PR for it.

**Comment by bangerth** [comment]
On 7/20/21 10:02 AM, Rene Gassmoeller wrote:
> @bangerth <https://github.com/bangerth>: I see where you are coming from, but 
> as @anne-glerum <https://github.com/anne-glerum> mentioned this is already 
> established (#3457 <https://github.com/geodynamics/aspect/pull/3457>), this PR 
> only uses the functionality in more places. I see why you say it could go into 
> the outputs object, but I think when I implemented this I had the following 
> two thoughts:
> 
>  1. Telling the material model what to do is an input operation.
>  2. Having this in the output object would mean it is handed over as
>     non-const, meaning the material model could modify what properties to
>     compute internally (not likely, but it seems conceptually not ok).
> 
> If you want it to be moved I don't feel very strongly about it, but my 
> preference would be an input. A fix would only need a bit of search and replace.

No need. I had forgotten that we already had that and not read the patch to 
remind myself of it.

> In any case I think we should not delay this PR for it.

Correct!

File: source/postprocess/visualization/seismic_anomalies.cc
Comment (anne-glerum) [reply source/postprocess/visualization/seismic_anomalies.cc:155]: I've added the property `additional_outputs` and included it in `all_properties` as well.

File: source/material_model/interface.cc
Comment (gassmoeller) [inline source/material_model/interface.cc:382]: This way you overwrite the existing value of requested_properties, which is different from the behavior in the if-branch. I think you were looking for only disabling viscosity? If so that would probably work like this:
```suggestion
        requested_properties = requested_properties & ~MaterialProperties::viscosity;
```

**Review by gassmoeller — COMMENTED** [review]
This looks nearly ready, but it acquired a few conflicts. If you make the one change I mentioned and rebase this should be good to go.

File: source/material_model/interface.cc
Comment (anne-glerum) [reply source/material_model/interface.cc:382]: Ah didn't know about `~`, cool!

**Comment by anne-glerum** [comment]
Fixed the merge conflicts and addressed your comment. 
This PR has become more than I originally described. The next step would be to remove the `compute_strainrate` flag altogether, in favor of the requested_properties, either using the constructor or the current form.

**Comment by anne-glerum** [comment]
Hmm, one test still fails. I'll look into it some more.

**Comment by anne-glerum** [comment]
I cannot seem to figure out why this last test shear_heating_with_melt is failing. @jdannberg would you maybe have an idea?

**Comment by bangerth** [comment]
@anne-glerum Would you mind rebasing this one more time to resolve the merge conflict?

**Comment by jdannberg** [comment]
I had a look at the shear_heating_with_melt test. It looks like the changes in the statistics file are very small. Basically, only the last (8th) digit changes, which really could be due to anything. It looks like the biggest changes are in the screen output in the residual of the compositional fields in the last nonlinear iteration of a given time step (so when the residuals are 1e-16, they might be twice than that or half of that).

The test uses the melt simple material model, which is one of the ones you changed, but I honestly do not see how your change could have affected the results (at least the change makes sense to me). In addition, there are other tests that use the same material model and also use operator splitting that have not changed at all. So I would say it's probably fine; at least I do not have any other good ideas.

**Comment by tjhei** [comment]
@zjiaqi2018 Do you have an idea what is going on, why the test output changed? Can you test if on current master tests/gmg_mesh_deform_ghost_entries.prm returns for ``RMS, max velocity`` with and without GMG?

**Comment by tjhei** [comment]
I talked with @zjiaqi2018 and it seems this prm might be problematic because we have seen free surface + GMG + elasticity to give incorrect solutions (it is in fact currently disabled, but not with a prescribed surface like this test).
That doesn't explain why the output changes here...

**Comment by tjhei** [comment]
@anne-glerum I asked @zjiaqi2018 to either find the problem or work around it by disabling/deleting the test temporarily so we can merge this PR.

@zjiaqi2018 It would still be good to know why the changes in this PR change the output. That doesn't make sense to me. Can you maybe confirm that RMS changes with this PR (try rebasing to current master to exclude the possibility that some unrelated change causes the issue).

**Comment by zjiaqi2018** [comment]
I ran gmg_mesh_deform_ghost_entries.prm with the deal.ii master, but got nan in `in.strain_rate`, I don't know what caused the problem.

**Comment by anne-glerum** [comment]
I've made some changes in stokes_matrix_free.cc, but they actually shouldn't change anything, because the defaults should be correct. But let's see what the tester says. 

@zjiaqi2018 Where/when did you get in.strain_rate = nan? Do you have a line nr or function name or something?

**Comment by zjiaqi2018** [comment]
> I've made some changes in stokes_matrix_free.cc, but they actually shouldn't change anything, because the defaults should be correct. But let's see what the tester says.
> 
> @zjiaqi2018 Where/when did you get in.strain_rate = nan? Do you have a line nr or function name or something?

image

image

**Comment by anne-glerum** [comment]
Thank you @zjiaqi2018 ! I suspect that in assembly.cc line 633 `need_viscosity` and line 578-579 `update_strain_rate` are not set correctly for the GMG case. Could you have a look and see whether they should be true in the case of failing test? I will set them to true in the code just to see if the test then succeeds.

**Comment by anne-glerum** [comment]
So was this test wrong before this PR? The test fails, but this branch gives the same velocities as the AMG:
```
-   Solving Stokes system... 34+0 iterations.
+   Solving Stokes system... 43+0 iterations.
 
-     RMS, max velocity:         0.142 m/year, 0.448 m/year
+     RMS, max velocity:         0.129 m/year, 0.408 m/year
```

**Comment by anne-glerum** [comment]
@tjhei Is the gmg_mesh_deform_ghost_entries.prm test broken? The GMG result isn't the same as for AMG on main (see my message from 20 days ago), but with this PR it is the same (and therefore different from the test results).

**Comment by tjhei** [comment]
Could be. @zjiaqi2018 ?

**Comment by tjhei** [comment]
Thank you, @anne-glerum this helped! Let me see if I can fix this correctly.

**Comment by tjhei** [comment]
@anne-glerum I think you can drop the last commit after you rebase when #4713 is merged.

**Comment by bangerth** [comment]
@tjhei suggested merging this before #4725. What is the status of this patch?

**Comment by tjhei** [comment]
Needs a rebase and the last commit to be dropped. Then some code reviews as well. At least I focused on the GMG bugs so far.

**Comment by anne-glerum** [comment]
I'm done presenting, so I can get back to my PR's. I'll rebase and drop the test commits today or tomorrow. And yes, then it needs a thorough review :)

**Review by tjhei — APPROVED** [review]
Great! Can you please squash your commits?

**Comment by anne-glerum** [comment]
@tjhei Anything else I should do?

File: source/postprocess/visualization/shear_stress.cc
Comment (tjhei) [reply source/postprocess/visualization/shear_stress.cc:62]: Yes, that would be a great new variant of reinit() after deprecating the old.

**Review by tjhei — COMMENTED** [review]
This all looks reasonable to me, but I think another pair of eyes would be good.

File: include/aspect/material_model/interface.h
Comment (gassmoeller) [inline include/aspect/material_model/interface.h:207]: I am not so sure about making this one separate from `additional_outputs`. Structurally it is a subset of `additional_outputs`. I understand we may want to ask for it separately, but in the current form there is a problem: Someone could ask for all `additional_outputs` and then assume all additional outputs will be computed, while `force_terms` will not be computed if not requested. Is there a reason we need this as a separate entry?

File: source/material_model/interface.cc
Comment (gassmoeller) [reply source/material_model/interface.cc:291]: Yes, I think this line is not needed.

**Review by gassmoeller — APPROVED** [review]
I would say let's get this merged and handle the few small comments in a follow-up. Because this PR touches so many files the risk of adding new conflicts soon is pretty large.

**Comment by gassmoeller** [comment]
And thanks for all the work @anne-glerum!

**Pull request merged**
